### ISSUE TITLE
[rush] Verify installation before running "build"

### ISSUE THREAD

**Issue by raihle** [microsoft/rushstack#2710]
## Summary

Rush should check whether dependencies are up-to-date before running `build` or other commands.

## Details

`rush build` (and many custom commands) typically requires that the installation is up-to-date in order to work as expected.

Cloning, checking out, or merging branches can easily lead to a situation where a developer forgets to run `rush install` or `rush update`, such that the build will either fail or otherwise produce a different result than it would in CI / on another machine.

As a user, I see two ways this can go:

* Check whether the installation and linking of dependencies is up-to-date when `rush build`, another bulk command, or a local (`rushx`) command is invoked
* If it is not up-to-date, either:
  * A) Run `rush install` automatically, or
  * B) Print a message and exit, e.g.:  
    > The installed dependencies do not match the requested dependencies. Please run `rush install` or `rush update` before re-running your command. If you know what you are doing, you can skip this check with the `--skip-install-check` flag.

In the long run it may be useful to configure whether each command requires install or not, but personally I would be more than happy with a command-line opt-out as above.

## Standard questions

Please answer these questions to help us investigate your issue more quickly:

| Question | Answer |
| -------- | -------- |
| `@microsoft/rush` globally installed version? | 5.46.1 |
| `rushVersion` from rush.json? | 5.46.1 |
| `useWorkspaces` from rush.json? | true |
| Operating system? | Windows |
| Would you consider contributing a PR? | Yes |
| Node.js version (`node -v`)? | 14.15.0 |

**Comment by iclanton** [comment]
This is a good idea, and it probably wouldn't be too difficult to implement. @raihle, would you be interested in putting this feature together?

**Comment by octogonz** [comment]
> If it is not up-to-date, either:
> 
>  * A) Run `rush install` automatically

Having `rush build` perform `rush install` automatically might be a step too far.  Currently Rush's docs suggest to use `rush update` for everything (because it is a superset of `rush install`), however at my work I have realized this is very confusing for people.  It's important for them to be aware of when they are regenerating their **pnpm-lock.yaml**.  The concepts of `rush install`, `rush update`, and `rush build` are pretty important, so it seems useful to separate them. I was going to change the docs to present it this way.

Then the **more steps = more clarity** workflow could be like:
``` 
$ rush build
Oops, you forgot to install your packages!  You need to do "rush install" first.

$ rush install
But hold on, your package.json file has changes that that impact your lockfile
-- for this you need "rush update""

$ rush update
$ rush build  # okay, now we're good
```

For the advanced user who already understands the workflow, it would be fine to provide something like:
```shell
# Shorthand for "rush install && rush build"
$ rush build --with-install
```

And for the advanced user who understands the implications of skipping the install, we could also provide something like:
```shell
# Suppress the error about me forgetting to run "rush install"
$ rush build --bypass-install
```

**Comment by raihle** [comment]
@iclanton Yeah, I'll try.  (I'll stick to checking & warning rather than automatically installing.)

The `BaseInstallManager` (+ derived classes) have a protected function `canSkipInstall` which seems useful. However, with pnpm workspaces enabled, the `WorkspaceInstallManager` also looks at every project's package.json, and these can change for many unrelated reasons. I see the check was added by @D4N14L. Do you know why these have to be checked? Maybe they can be skipped for this case? 

Maybe a more conservative check (just shrinkwrap-file vs common/temp's LastInstallFlag) would be better? I'd rather have false negatives than false positives in this case.

**Comment by D4N14L** [comment]
Main reason package.json is being checked is because legacy Rush installs look at tarballs, which are empty save for the package.json. So you can use the package.json as an analog for this tarball. As for the reasoning for looking at _those_ files, it's important to do so to a) ensure that no versions in the package.json have been modified since the last install, and because NPM has `preinstall` and `postinstall` commands that can run before or after install (respectively) so they also need to be taken into account.

It's important to keep in mind that `canSkipInstall` is only one method used to determine if install should be skipped/is up to date, and there are also other factors. For example, `prepareCommonTempAsync` returns `shrinkwrapUpToDate` and `variantUpToDate`, since we are already doing preparation of the common/temp folder here. So you'd need to separate that out and perform that check elsewhere (maybe in the `canSkipInstall` call?)
